### ISSUE TITLE
module.register_full_backward_hook doesn't work

### ISSUE THREAD

**Issue by guotuofeng** [pytorch/pytorch#54405]
## 🐛 Bug

the backward hook registered by Module.register_full_backward_hook doesn't work. But the same hook works when registering using torch.nn.modules.module.register_module_full_backward_hook.

## To Reproduce

```

import sys
import torch
import torchvision
import torchvision.transforms as transforms
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim

import time

class MyModule(nn.Module):
    def __init__(self):
        super(MyModule, self).__init__()
        self.conv1 = nn.Conv2d(3, 6, 5)
        self.pool = nn.MaxPool2d(2, 2)

    def forward(self, x):
        x = self.pool(F.relu(self.conv1(x)))
        return x

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.sb = MyModule()
        self.pool = nn.MaxPool2d(2, 2)
        self.conv2 = nn.Conv2d(6, 16, 5)
        self.fc1 = nn.Linear(16 * 5 * 5, 120)
        self.fc2 = nn.Linear(120, 84)
        self.fc3 = nn.Linear(84, 10)

        def backward_hook(module, grad_input, grad_output):
            print("backward_hook on module", id(module))

        #torch.nn.modules.module.register_module_full_backward_hook(backward_hook)
        self.sb.register_full_backward_hook(backward_hook)

    def forward(self, x):
        x = self.sb(x)
        x = self.pool(F.relu(self.conv2(x)))
        x = x.view(-1, 16 * 5 * 5)
        x = F.relu(self.fc1(x))
        x = F.relu(self.fc2(x))
        x = self.fc3(x)
        return x

if __name__ == '__main__':
    print("{}, argv={}".format("test", sys.argv))

    torch.manual_seed(0)

    transform = transforms.Compose(
        [transforms.ToTensor(),
         transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])

    trainset = torchvision.datasets.CIFAR10(root='./data', train=False,
                                            download=True, transform=transform)
    trainloader = torch.utils.data.DataLoader(trainset, batch_size=4,
                                              shuffle=True, num_workers=0)

    model = Net()
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

    for i, data in enumerate(trainloader):
        inputs, labels = data
        optimizer.zero_grad()
        outputs = model(inputs)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()

        print('[%d] loss: %.3f' % (i + 1, loss.item()))
        time.sleep(0.5)
        if i == 10:
            break

```

## Expected behavior

The backward hook is called.

## Environment
Collecting environment information...
PyTorch version: 1.8.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Enterprise
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.19.4

Python version: 3.7 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Quadro K620
GPU 1: Quadro K620

Nvidia driver version: 461.40
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.0
[pip3] torchvision==0.9.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.0                   pypi_0    pypi
[conda] mkl-include               2019.0                   pypi_0    pypi
[conda] mkl-service               2.3.0            py37h196d8e1_0
[conda] mkl_fft                   1.2.0            py37h45dec08_0
[conda] mkl_random                1.1.1            py37h47e9c7a_0
[conda] numpy                     1.19.1                   pypi_0    pypi
[conda] numpy-base                1.19.2           py37ha3acd2a_0
[conda] torch                     1.8.0                    pypi_0    pypi
[conda] torchvision               0.9.0                    pypi_0    pypi

## Additional context

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

**Comment by albanD** [comment]
Hi,

I am afraid this is expected.
The inputs to the Module does not require gradients, so there is nothing to call the hook with.
You can add `inputs.requires_grad_()` to make sure it gets called properly.

**Comment by albanD** [comment]
Maybe we want to add a warning [here](https://github.com/pytorch/pytorch/blob/36ce673f16269724f629f2a1897b8720ba6c8f18/torch/nn/modules/module.py#L1033) to tell the users to make this less confusing.

**Comment by guotuofeng** [comment]
But the global backward hook and original register_backward_hook works. 
Are the inconsistent behavior expected?

**Comment by albanD** [comment]
The global hooks will fire on all Modules. SO in this case they do fire for the other nn.Modules in your model.
The old style hooks returns wrong values (they have to because what you ask for does not exist here). So it is expected that the behavior is not the same.

**Comment by guotuofeng** [comment]
thanks for the clarification
